### ISSUE TITLE
support FSDP with AMP

### ISSUE THREAD

**Issue by rosrad** [pytorch/pytorch#76607]
### 🚀 The feature, motivation and pitch

Replacing the DDP with FSDP results into a loss difference, where the AMP is enabled.

Do we have some document to help users enable both FSDP and AMP ?

Currently simply changing from DDP to FSDP can lead to some loss regression as below.
![image](https://user-images.githubusercontent.com/2343640/166081973-a0099713-3be0-4ff7-942b-cbb3eb661158.png)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by cbalioglu** [comment]
cc @zhaojuanmao @rohan-varma

**Comment by rohan-varma** [comment]
As @zhaojuanmao mentioned, we're building out a shard-aware gradient scaler similar to `torch.cuda.amp.GradScaler` to help with gradient scaling when training with AMP.

In addition, we have added mixed precision training with FSDP with https://github.com/pytorch/pytorch/pull/75024 that can also be used for mixed precision with FSDP. For fp16 training, both will likely require sharded_grad_scaler, although mixed precision training can be tried with `bfloat16` if there is access to A100 which has shown similar speedup with less (minimal) accuracy regression.

To confirm, are you using Gradient Scaling with DDP + torch.cuda.amp?

**Comment by rosrad** [comment]
Yes, currently we are using the gradient scaling with DDP+amp. 
By the way, is that already supported of FSDP with bfloat16? 
If this is the case, can you share some examples how to use bfloat16 with FSDP or DDP

**Comment by daniellepintz** [comment]
Hi, I am wondering what is the status of sharded_grad_scaler?

**Comment by eric-haibin-lin** [comment]
@zhaojuanmao is the sharded grad scaler available now?

**Comment by chrjxj** [comment]
> Yes, currently we are using the gradient scaling with DDP+amp. By the way, is that already supported of FSDP with bfloat16? If this is the case, can you share some examples how to use bfloat16 with FSDP or DDP

@zhaojuanmao @rohan-varma   hi, have the same question as above. 

other than example in https://pytorch.org/tutorials/intermediate/FSDP_adavnced_tutorial.html ,  haven't found other example/cases using FSDP+AMP (bf16)

**Comment by zhaojuanmao** [comment]
yes, sharded_grad_scaler is available now. see https://github.com/pytorch/pytorch/blob/master/torch/distributed/fsdp/sharded_grad_scaler.py, it has some simple examples in this test file https://github.com/pytorch/pytorch/blob/master/test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py

To use FSDP + bf16, you can simply wrapping FSDP like this, it should work with or without AMP: 

fsdp_model = FSDP(...., mixed_precision= MixedPrecision(param_dtype=bfloat16, reduce_dtype=bfloat16))

**Comment by david-macleod** [comment]
@zhaojuanmao is using the sharded_grad_scaler the recommended approach to training with FSDP + autocast(torch.fp16) or still experimental? 

I ask because it is undocumented, for example does not appear in the recipes here https://pytorch.org/docs/stable/notes/amp_examples.html#working-with-multiple-gpus, so was wondering if this is just a less common setup and people are more likely to to use FSDP + autocast(torch.bfloat16) without any grad_scaler?

The only reference I can see is this broken link https://pytorch.org/docs/main/distributed.html#module-torch.distributed.fsdp.sharded_grad_scaler

**Comment by awgu** [comment]
@david-macleod  People most commonly use FSDP's own built-in mixed precision (`mixed_precision: MixedPrecision` arg) because it offers more memory savings and can be faster. The difference is that FSDP's mixed precision does not cast to bf16/fp16 per op but rather holds the parameters and activations in bf16/fp16 during forward/backward computation and FSDP's mixed precision frees the bf16/fp16 copy to avoid holding it through forward/backward.

If you use FSDP's built-in mixed precision with fp16, then you should also use `ShardedGradScaler`. If using bf16, then there is no need.

**Comment by david-macleod** [comment]
great, thanks!

**Comment by samsja** [comment]
> @david-macleod People most commonly use FSDP's own built-in mixed precision (`mixed_precision: MixedPrecision` arg) because it offers more memory savings and can be faster. The difference is that FSDP's mixed precision does not cast to bf16/fp16 per op but rather holds the parameters and activations in bf16/fp16 during forward/backward computation and FSDP's mixed precision frees the bf16/fp16 copy to avoid holding it through forward/backward.
> 
> If you use FSDP's built-in mixed precision with fp16, then you should also use `ShardedGradScaler`. If using bf16, then there is no need.

would be worth to add to the docs
